sbt-plugin/src/main/scala-sbt-1.0/org/scalajs/sbtplugin/SJSIRFileManager.scala

```diff
@@ -0,0 +1,100 @@
+/*
+ * Scala.js (https://www.scala-js.org/)
+ *
+ * Copyright EPFL.
+ *
+ * Licensed under Apache License 2.0
+ * (https://www.apache.org/licenses/LICENSE-2.0).
+ *
+ * See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.
+ */
+
+package org.scalajs.sbtplugin
+
+import java.io.File
+import java.nio.file.Files
+
+import scala.collection.mutable
+
+import sbt._
+
+import xsbti.compile.ClassFileManager
+
+/** A class file manager that prunes .sjsir files as needed.
+ *
+ *  This makes sure that, when a .class file must be deleted, the
+ *  corresponding .sjsir file is also deleted, as well as maintaining the
+ *  transactional aspect of incremental compilation in the presence of
+ *  compilation errors in an intermediate run (as product files need to be
+ *  backed up and later restored).
+ *
+ *  This code is adapted from Zinc `TransactionalClassFileManager`.
+ *  We need to duplicate the logic since forwarding to the default class
+ *  file manager doesn't work: we need to backup sjsir files in a different
+ *  temporary directory as class files.
+ *
+ *  We stole this from `TastyFileManager` in dotty, at
+ *  https://github.com/lampepfl/dotty/blob/0.17.0-RC1/sbt-dotty/src/dotty/tools/sbtplugin/TastyFileManager.scala
+ *
+ *  There are plans to upstream a generic version of this code in zinc in the
+ *  future: https://github.com/sbt/zinc/issues/579
+ */
+private[sbtplugin] final class SJSIRFileManager extends ClassFileManager {
+  private[this] var _tempDir: File = null
+
+  private[this] def tempDir: File = {
+    if (_tempDir == null) {
+      _tempDir = Files.createTempDirectory("backup").toFile
+    }
+    _tempDir
+  }
+
+  private[this] val generatedSJSIRFiles = new mutable.HashSet[File]
+  private[this] val movedSJSIRFiles = new mutable.HashMap[File, File]
+
+  override def delete(classes: Array[File]): Unit = {
+    val files = sjsirFiles(classes)
+    val toBeBackedUp = files
+      .filter(t => t.exists && !movedSJSIRFiles.contains(t) && !generatedSJSIRFiles(t))
+    for (c <- toBeBackedUp)
+      movedSJSIRFiles.put(c, backupFile(c))
+    IO.deleteFilesEmptyDirs(files)
+  }
+
+  override def generated(classes: Array[File]): Unit =
+    generatedSJSIRFiles ++= sjsirFiles(classes)
+
+  override def complete(success: Boolean): Unit = {
+    if (!success) {
+      IO.deleteFilesEmptyDirs(generatedSJSIRFiles)
+      for ((orig, tmp) <- movedSJSIRFiles)
+        IO.move(tmp, orig)
+    }
+
+    generatedSJSIRFiles.clear()
+    movedSJSIRFiles.clear()
+    if (_tempDir != null) {
+      IO.delete(tempDir)
```

Comment:
This will create the tempDir again. Maybe name it differently so it is clear it creates it?

Reply:
Hum, we just tested that it is not null, so it will not recreate it. It will only delete the one that was already created.

We could use `_tempDir` to make it clearer though.